<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2015-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Joel Gómez Uribe
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 1.  Programación para la física computacional


### Programación básica para la física

---

### EJERCICIOS:


**1. Órbitas planetarias:** En el espacio, la órbita de unc uerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancias $\ell_{1}$ de máxima aproximación de un planeta al Sol (su _perihelio_), y su velocidad inicial $v_{1}$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

(a) La segunda ley de Kepler nos dice que la distancia $\ell_{2}$ y la velocidad $v_{2}$ del planeta en su punto más distante, o _afelio_, satisfacen que $\ell_{2}v_{2} = \ell_{1}v_{1}$. Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

\begin{equation}
    E = \dfrac{1}{2}mv^{2} - G\dfrac{mM}{r},
\end{equation}

donde $m$ es la masa del planeta, $M = 1.9891\times 10^{30}$ kg es la masa del Sol y $6.6738\times 10^{-11}$ m$^{3}$ kg$^{-1}$ s$^{-2}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_{2}$ es la raíz más pequeña de la ecuación cuadrática:

\begin{equation}
    v_{2}^{2} - \dfrac{2GM}{v_{1}\ell_{1}}v_{2} - \Biggl[v_{1}^{2} - \dfrac{2GM}{\ell_{1}}\Biggr] = 0.
\end{equation}

(b) Dados los valores de $v_{1},\ell_{1}$ y $\ell_{2}$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

\begin{align*}
    \text{Semieje mayor}:\quad a &= \tfrac{1}{2}(\ell_{1} + \ell_{2}),\\
    \text{Semieje menor}:\quad b &= \sqrt{\ell_{1}\ell_{2}},\\
    \text{Período orbital}:\quad T &= \dfrac{2\pi ab}{\ell_{1}v_{1}},\\
    \text{Excentricidad orbital}:\quad e &= \dfrac{\ell_{2} - \ell_{1}}{\ell_{2} + \ell_{1}}.
\end{align*}

Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velcidad en el perihelio; para que calcule e imprima las cantitades $\ell_{2},v_{2}, T$ y $e$.

(c) Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_{1} = 1.4710\times 10^{11}$m y $v_{1} = 3.0287\times 10^{4}$ms$^{-1}$) y del cometa Halley ($\ell_{1} = 8.7830\times 10^{10}$m y $v_{1} = 5.4529\times 10^{4}$ms$^{-1}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley de unos 76 años.

**2. La constante de Madelung**

En física de la materia condensada, la _constante de Madelung_ da el potencial eléctrico total que siente un átomo en un sólido; y depende de las cargas de los otros átomos cercanos y de sus ubicaciones.

Por ejemplo, el cristal de cloruro de sodio sólido (la sal de mesa), tiene átomos dispuestos en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio una carga positiva $+e$ y los de cloro una negativa $-e$, (donde $e$ es la carga del electrón). Si etiquetamos cada posición en la red con tres coordenadas enteras $(i,j,k)$, entonces los átomos de sodio caen en posiciones donde $i + j + k$ es para, y los átomos de clore en posiciones donde $i + j + k$ es impar.

Consideremos un átomo de sodio en el origen, i.e. $i = j = k =0$, y calculemos la _constante de Madelung_. Si el espacio de los átomos en la red es $a$, entonces la distancia desde el origen al átomo en la posición $(i,j,k)$ es:

\begin{equation}
    \sqrt{(ia)^{2} + (ja)^{2} + (ka)^{2}} = a \sqrt{i^{2} + j^{2} + k^{2}},
\end{equation}

y el potencial en el origen creado por el átomo es:

\begin{equation}
    V(i,j,k) = \pm \dfrac{e}{4\pi\varepsilon_{0}a\sqrt{i^{2} + j^{2} + k^{2}}},
\end{equation}

siendo $\varepsilon_{0}$ la permitividad del vacío y el signo de la expresión se toma dependiendo de si $i + j + k$ es par o impar. Así entonces, el potencial total que siente el átomo de sodio es la suma de esta cantidad sobre los demás átomos. Supongamos una caja cúbica alrededor del átomo de sodio en el origen, con $L$ átomos en todas la direcciones, entonces:

\begin{equation}
    V_{\text{total}} = \sum\limits_{\substack{i,j,k=-L\\i,j,k\neq 0}}^{L} V(i,j,k) = \dfrac{e}{4\pi\varepsilon_{0}a}M,
\end{equation}

donde $M$ es la constante de Madelung (al menos aproximadamente).

Técnicamente, la constante de Madelung es el valor de $M$ cuando $L\to\infty$, pero se puede obtener una buena aproximación simplemente usando un valor grande de $L$.
Escribe un programa para calcular e imprimir la _constante de Madelung_ para el cloruro de sodio. Utiliza un valor de $L$ tan grande como puedas, sin dejar que tu programa se ejecute en un tiempo razonable (un minuto o menos).

In [31]:
def madelung(L):
    M = 0
    
    for i in range(-L, L + 1):
        for j in range(-L, L + 1):
            for k in range(-L, L + 1):

                if i == j == k == 0:
                    continue

                potencial = ((i**2 + j**2 + k**2) ** (1 / 2))

                if (i + j + k) % 2 == 0:
                    M += 1 / potencial
                else:
                    M -= 1 / potencial
    return M

In [32]:
# %%time
madelung(300)

-1.745643295912666

[Constante de Madelung](https://en.wikipedia.org/wiki/Madelung_constant)

**3. Coeficientes binomiales**

El coeficiente binomial $\begin{pmatrix}n\\k\end{pmatrix}$ es un número entero igual a:

\begin{equation}
    \begin{pmatrix}n\\k\end{pmatrix} = \dfrac{n!}{k!(n - k)!} = \dfrac{n\times(n - 1)\times\cdots\times(n - k + 1)}{1\times 2\times \cdots\times k},
\end{equation}

donde $k\leq 1$, o bien $\begin{pmatrix}n\\0\end{pmatrix} = 1$ cuando $k = 0$.

(a) Utiliza esta fórmula para escribir una función llamada ``binomial(n,k)`` (o como tú quieras) que calcule el coeficiente binomial para $n$ y $k$ dados. Asegúrate de que tu función devuelva la respuesta en forma de número entero (no flotante) y proporcione el valor correcto de 1 para el caso en que $k = 0$.

In [33]:
def factorial(n:int) -> int:
    if n < 2:
        return 1
    else:
        return n * factorial(n - 1)


def binomial(n: int, k: int) -> int:
    return factorial(n) // (factorial(k) * factorial(n - k))

In [34]:
for i in range(5):
    print(binomial(i, 0))

1
1
1
1
1


(b) Usando tu función, escribe un programa que imprima las primera 20 líneas del "_triángulo de Pascal_". La $n$-ésima línea del triángulo de Pascal continene $n + 1$ números, que son los coeficientes $\begin{pmatrix}n\\0\end{pmatrix}$, $\begin{pmatrix}n\\1\end{pmatrix}$, y así sucesivamente $\begin{pmatrix}n\\k\end{pmatrix}$. De tal manera que las primeras líneas son:

**1**  
**1 1**  
**1 2 1**  
**1 3 3 1**  
**1 4 6 4 1**

In [35]:
def triangulo_pascal(lineas: int):
    for n in range(lineas):
        for k in range(n + 1):
            print(f" {binomial(n, k)}", end=" ")
    
        print()

In [36]:
triangulo_pascal(20)

 1 
 1  1 
 1  2  1 
 1  3  3  1 
 1  4  6  4  1 
 1  5  10  10  5  1 
 1  6  15  20  15  6  1 
 1  7  21  35  35  21  7  1 
 1  8  28  56  70  56  28  8  1 
 1  9  36  84  126  126  84  36  9  1 
 1  10  45  120  210  252  210  120  45  10  1 
 1  11  55  165  330  462  462  330  165  55  11  1 
 1  12  66  220  495  792  924  792  495  220  66  12  1 
 1  13  78  286  715  1287  1716  1716  1287  715  286  78  13  1 
 1  14  91  364  1001  2002  3003  3432  3003  2002  1001  364  91  14  1 
 1  15  105  455  1365  3003  5005  6435  6435  5005  3003  1365  455  105  15  1 
 1  16  120  560  1820  4368  8008  11440  12870  11440  8008  4368  1820  560  120  16  1 
 1  17  136  680  2380  6188  12376  19448  24310  24310  19448  12376  6188  2380  680  136  17  1 
 1  18  153  816  3060  8568  18564  31824  43758  48620  43758  31824  18564  8568  3060  816  153  18  1 
 1  19  171  969  3876  11628  27132  50388  75582  92378  92378  75582  50388  27132  11628  3876  969  171  19  1 


(c) La probabilidad de que para una moneda no sesgada, lanzada $n$ veces, salga águila $k$ veces es:

\begin{equation}
    p(k | n) = \dfrac{\begin{pmatrix}n\\k\end{pmatrix}}{2^{n}}.
\end{equation}

Escribe un programa para calcular:

i. la probabilidad total de que un a moneda lanzada 100 veces, salga águila exactamente 60 veces

In [42]:
# def probabilidad(k: int, n: int):
def probabilidad_total(n: int, k: int):
    prob_tot = binomial(n, k) / (2 ** n)
    
    return n,k,prob_tot

aguila_60 = probabilidad_total(100, 60)
print(f"La probabilidad total después de lanzar una moneda {aguila_60[0]} veces y que salga exactamente {aguila_60[1]} veces águila es de {aguila_60[2]}")

La probabilidad total después de lanzar una moneda 100 veces y que salga exactamente 60 veces águila es de 0.010843866711637987


ii. la probabilidad de que salga águila 60 veces o más.

In [43]:
def probabilidad_mas(n: int, k: int):
    pp = 0
    
    for i in range(k, n + 1):
        pp += binomial(n, i) / (2 ** n)

    return n, k, pp

aguila_mas_60 = probabilidad_mas(100, 60)

print(f"La probabilidad de obtener águila más de {aguila_mas_60[1]} veces es de {aguila_mas_60[2]}.")

La probabilidad de obtener águila más de 60 veces es de 0.028443966820490392.


**4. Números primos**

Es posible escribir un programa bastante rápido para calcular número sprimos siguiendo las siguientes observaciones:

(a) Un número $n$ es primo, si no tiene factores primos menores que $n$. Por lo tanto, solo necesitamos comprobar si es divisible por otros números primos.  
(b) Si un número $n$ no es primo y tiene un factor $r$, entonces $n = rs$, donde $s$ también es un factor. Si $r \leq \sqrt{n}$ entonces $n = rs \leq \sqrt{n}s$, lo que implica que $s \geq \sqrt{n}$. Es decir, cualquier **no primo** debe tener factores, y por lo tanto también factores primos, menores o iguales a $\sqrt{n}$. POr lo tanto, para determinar si un número es primo, tenemos que verificar sus factores primos solo hasta $\sqrt{n}$; incluso si no hay ninguno, entonces el número es primo.  
(c) Si no encontramos un **factor primo menor que $\sqrt{n}$ entonces sabemos que el número no es primo** y, por lo tanto, no hay necesidad de realizar más comprobaciones; podemos abandonar este número y pasar al siguiente.

**Escribe un programa que encuentre todos los números primos hasta $10000$ (diez mil)**. Crea una lista para almacenar los números primos, que comience con un solo número primo (el 2). Luego, para cada número $n$ del $3$ al $10000$, verifica si el número es divisible por alguno de los números primos de la lista hasta $\sqrt{n}$. En cuanto encuentres un factor primo, puedes dejar de comprobar el resto de número (pues ya sabrías que $n$ no es primo). Si no encuentras factores primos iguales o menores a $\sqrt{n}$, entonces $n$ es primo y habría que agregarlo a la lista. Puedes imprimir la lista completa al final del programa, o bien, ir imprimiendo los números indivualmente a medida que los vayas encontrando.

In [28]:
def primos(n: int):    
    nums_primos = [2]
    
    for num in range(3, n + 1):
        for i in range(2, int(num ** (1 / 2)) + 1):
            if num % i == 0:
                break
        else:
            nums_primos.append(num)
    return nums_primos

In [30]:
print(primos(10000))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 12

**5.Recursión**

Una característica muy útil de las funciones que podemos definir es la **_recursividad_**, es decir, la capacidad de una función de llamarse a sí misma. Por ejemplo, considere la siguiente definición del factorial $n!$ de un entero positivo $n$:

\begin{equation}
    n! = 
    \begin{cases}
        1 & \text{si}\; n = 1,\\
        n\times(n - 1)! & \text{si}\; n > 1.
    \end{cases}
\end{equation}

Esto constituye una definición completa del factorial que nos permite calcular el valor de $n!$ para cualquier otro entero positivo. Podemos emplear esta definición directamente para crear una función para calcular factoriales, como esta (en Python):

```python
def factorial(n):
    if n == 1:
        return 1
    else:
        return n * factorial(n - 1)
```

Nota cómo si $n$ no es igual a 1, la función se llama a sí misma para calcular el factorial de $n - 1$, de tal manera que al pedir "print(factorial(5))" el intérprete de Python imprimirá correctamente la respuesta 120.

(a) Anteriormente, en clase encontramos los número de Catalán $C_{n}$. Si hacemos una pequeña modificación, la definición de $C_{n}$ se puede reescribir de la siguiente forma:

\begin{equation}
    C_{n} = 
    \begin{cases}
        1\quad & \text{if}\; n = 0,\\
        \dfrac{4n - 2}{n + 1}C_{n - 1}\quad & \text{if}\; n > 0,
    \end{cases}
\end{equation}

Escribe una función, **usando recursividad**, que calcule $C_{n}$ y utilízala para calcular e imprimir C_{100}.

In [2]:
def catalan(n: int):
    if n == 0:
        return 1
    else:
        op = ((4 * n) - 2) / (n + 1)
        return op * catalan(n - 1)

In [2]:
catalan(0)

1

In [3]:
catalan(100)

8.965199470901317e+56

(b) Euclides demostró que el máximo común divisor $g(m,n)$ de dos enteros no negativos $m$ y $n$ satisface que:

\begin{equation}
    g(m,n) =
    \begin{cases}
        m\quad & \text{si}\; n = 0,\\
        g(n, m\ \text{mod}\ n)\quad & \text{si}\; n > 0.
    \end{cases}
\end{equation}

Usando esta fórmula escribe una función ``g(m, n)`` que emplee **recursividad** para calcular el máximo común divisor de $m$ y $n$. Usa tu función para calcular e imprimir el máximo común divisor de 108 y 192.

In [4]:
def g(m: int, n: int):
    if n == 0:
        return m
    else:
        return g(n, m % n)

In [6]:
print(f"El máximo común divisor de 108 y 192 es {g(108, 192)}.")

El máximo común divisor de 108 y 192 es 12.


Comparando el cálculo de los números de Catalán del inciso 5a, con lo hecho en clase, vemos que es posible hacer el cálculo de dos maneras, ya sea directamente o usando _recursividad_.

En la mayoría de los casos, si una cantidad se puede calcular **_sin_ recursividad**, será más rápido y normalmente se recomienda hacerlo así, si es posible. Sin embargo, hay algunos cálculos que son esencialmente imposibles (o al menos mucho más difíciles) sin recursividad.